s3/src/s3_lib.py

```diff
@@ -0,0 +1,372 @@
+        # FIXME
+        if self.bucket is None:
+            self.bucket = f"relation-{event.relation.id}"
+        event_data = {"bucket": self.bucket}
+        self.relation_data.update_relation_data(event.relation.id, event_data)
+
+    def get_s3_connection_info(self) -> dict[str, str]:
+        """Return the S3 connection info as a dictionary."""
+        for relation in self.relations:
+            if relation and relation.app:
+                info = self.relation_data.fetch_relation_data([relation.id])[relation.id]
+                if set(S3_REQUIRED_OPTIONS) - set(info):
+                    continue
+                return info
+        return {}
+
+    def _on_relation_changed_event(self, event: RelationChangedEvent) -> None:
+        """Notify the charm about the presence of S3 credentials."""
+        logger.info(f"S3 relation ({event.relation.name}) changed...")
+
+        diff = self._diff(event)
+        if any(newval for newval in diff.added if self.relation_data._is_secret_field(newval)):
+            self.relation_data._register_secrets_to_relation(event.relation, diff.added)
+
+        # check if the mandatory options are in the relation data
+        contains_required_options = True
+        credentials = self.get_s3_connection_info()
+        missing_options = []
+        for configuration_option in S3_REQUIRED_OPTIONS:
+            if configuration_option not in credentials:
+                contains_required_options = False
+                missing_options.append(configuration_option)
+
+        # emit credential change event only if all mandatory fields are present
+        if contains_required_options:
+            getattr(self.on, "s3_connection_info_changed").emit(
+                event.relation, app=event.app, unit=event.unit
+            )
+        else:
+            logger.warning(
+                f"Some mandatory fields: {missing_options} are not present, do not emit credential change event!"
+            )
+
+    def _on_secret_changed_event(self, event: SecretChangedEvent) -> None:
+        """Event handler for handling a new value of a secret."""
+        if not event.secret.label:
+            return
+
+        relation = self.relation_data._relation_from_secret_label(event.secret.label)
+        if not relation:
+            logging.info(
+                f"Received secret {event.secret.label} but couldn't parse, seems irrelevant."
+            )
+            return
+
+        if event.secret.label != self.relation_data._generate_secret_label(
+            relation.name,
+            relation.id,
+            "extra",
+        ):
+            logging.info("Secret is not relevant for us.")
+            return
+
+        if relation.app == self.charm.app:
+            logging.info("Secret changed event ignored for Secret Owner")
+
+        remote_unit = None
+        for unit in relation.units:
+            if unit.app != self.charm.app:
+                remote_unit = unit
```

Comment:
- Nit: `break` after?
- An unlikely edge case, but should we add `else: return`? ([for-else](https://docs.python.org/3.8/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops))

Reply:
Agreed, done in 346c4b455123ed5f1757542d9fb3999fe80a0815.